packages/backend/src/services/RenameService/RenameService.ts

```diff
@@ -0,0 +1,695 @@
+        this.schedulerClient = args.schedulerClient;
+        this.schedulerModel = args.schedulerModel;
+    }
+
+    public async getFieldsForChart({
+        user,
+        projectUuid,
+        chartUuid,
+    }: {
+        user: SessionUser;
+        projectUuid: string;
+        chartUuid: string;
+    }) {
+        const chart = await this.savedChartModel.get(chartUuid);
+        const explore = await this.projectModel.getExploreFromCache(
+            projectUuid,
+            chart.tableName,
+        );
+        if (isExploreError(explore)) {
+            throw new NotFoundError(
+                `Valid explore ${chart.tableName} not found`,
+            );
+        }
+        // TODO also add table calculations, custom metrics and dimensions
+        // TODO also add joins
+        const fields = Object.keys(
+            explore.tables[explore.baseTable].dimensions,
+        ).concat(Object.keys(explore.tables[explore.baseTable].metrics));
+        return {
+            [explore.baseTable]: fields.map((f) =>
+                getItemId({ name: f, table: explore.baseTable }),
+            ), // add explore prefix
+            tableCalculations: chart.metricQuery.tableCalculations?.map(
+                (tc) => tc.name,
+            ),
+            customMetrics: chart.metricQuery.additionalMetrics?.map(
+                (tc) => tc.name,
+            ),
+            customDimensions: chart.metricQuery.customDimensions?.map(
+                (tc) => tc.name,
+            ),
+        };
+    }
+
+    /**
+     * Triggered from the UI on validation settings
+     */
+    async renameChart({
+        user,
+        projectUuid,
+        chartUuid,
+        from,
+        to,
+        type,
+        fixAll,
+        context,
+    }: ApiRenameChartBody & {
+        user: SessionUser;
+        projectUuid: string;
+        context: RequestMethod;
+    }) {
+        if (from === to) {
+            throw new ParameterError(
+                'Old and new names are the same, nothing to rename',
+            );
+        }
+        const { organizationUuid } = await this.projectModel.getSummary(
+            projectUuid,
+        );
+        if (
+            user.ability.cannot(
+                'update',
+                subject('Project', {
+                    organizationUuid,
+                    projectUuid,
+                }),
+            )
+        ) {
+            throw new ForbiddenError();
+        }
+
+        const chart = await this.savedChartModel.get(chartUuid);
+
+        const nameChanges = getNameChanges({
+            from,
+            to,
+            table: chart.tableName,
+            type,
+        });
+
+        if (type === RenameType.MODEL) {
+            // Verify the model we want to assign to really exists
+            // This will throw an error if the model does not exist
+            await this.projectModel.getExploreFromCache(projectUuid, to);
+        } else {
```

Comment:
switch case with `assertUnreachable` will make this typesafe and code readable since else case is not instantly obvious that it works on fields